src/meili/map_matcher.cc

```diff
@@ -113,74 +112,43 @@ InterpolateMeasurements(const MapMatching& mapmatching,
+  //we weren't able to get a downstream route
   std::vector<MatchResult> results;
-
-  if (!state.IsValid()) {
-    for (const auto& measurement: interpolated_measurements) {
+  std::vector<EdgeSegment> route;
+  if (!state.IsValid() || !next_state.IsValid() ||
+    MergeRoute(route, *state, *next_state).empty()) {
+    for (const auto& measurement: interpolated_measurements)
       results.emplace_back(MatchResult{measurement.lnglat(), 0.f, baldr::GraphId{}, -1.f, measurement.epoch_time(), kInvalidStateId});
-    }
-    return results;
-  }
-
-  std::vector<EdgeSegment> upstream_route;
-  if (previous_state.IsValid()) {
-    MergeRoute(upstream_route, *previous_state, *state);
-  }
-
-  std::vector<EdgeSegment> downstream_route;
-  if (next_state.IsValid()) {
-    MergeRoute(downstream_route, *state, *next_state);
-  }
-
-  if (upstream_route.empty() && downstream_route.empty()) {
-    for (const auto& measurement: interpolated_measurements) {
-      results.emplace_back(MatchResult{measurement.lnglat(), 0.f, baldr::GraphId{}, -1.f, measurement.epoch_time(), kInvalidStateId});
-    }
     return results;
   }
 
+  //for each point that needs interpolated
   for (const auto& measurement: interpolated_measurements) {
     const auto& match_measurement = mapmatching.measurement(state->time());
     const auto match_measurement_distance = GreatCircleDistance(measurement, match_measurement);
-
-    const auto& up_interp = InterpolateMeasurement(
-        mapmatching,
-        upstream_route.rbegin(),
-        upstream_route.rend(),
-        measurement,
-        match_measurement_distance,
-        true);
-
-    const auto& down_interp = InterpolateMeasurement(
-        mapmatching,
-        downstream_route.begin(),
-        downstream_route.end(),
-        measurement,
-        match_measurement_distance,
-        false);
-
-    if (up_interp.edgeid.Is_Valid() && down_interp.edgeid.Is_Valid()) {
-      const auto down_cost = down_interp.sortcost(mapmatching, match_measurement_distance),
-                   up_cost = up_interp.sortcost(mapmatching, match_measurement_distance);
-      if (down_cost < up_cost) {
-        results.emplace_back(MatchResult{down_interp.projected, std::sqrt(down_interp.sq_distance), down_interp.edgeid, down_interp.edge_distance, measurement.epoch_time(), kInvalidStateId});
-      } else {
-        results.emplace_back(MatchResult{up_interp.projected, std::sqrt(up_interp.sq_distance), up_interp.edgeid, up_interp.edge_distance, measurement.epoch_time(), kInvalidStateId});
+    //interpolate this point along the route
+    const auto& interp = InterpolateMeasurement(mapmatching, route.begin(), route.end(), measurement, match_measurement_distance);
+
+    //if it was able to do the interpolation
+    if (interp.edgeid.Is_Valid()) {
+      //dont allow subsequent points to get interpolated before this point
+      //we do this by editing the route to start where this point was interpolated
```

Comment:
@ptpt does this seem ok to you? are there ever any consecutive duplicate `edgeid`s within the `EdgeSegment`s in the `route`? if not then i can make this simpler.